text/3830-dedented-string-literals.md

```diff
@@ -0,0 +1,1296 @@
+
+    Before:
+
+    ```rs
+    eprintdoc! {"
+            GET {url}
+            Accept: {mime}
+        ",
+        ^^^^ // common leading whitespace (will be removed)
+        url = "http://localhost:8080",
+        mime = "application/json",
+    }
+    ```
+
+    With dedented string literals:
+
+    ```rs
+    eprintln! {
+        d"
+            GET {url}
+            Accept: {mime}
+        ",
+        ^^^^ // common leading whitespace (will be removed)
+        url = "http://localhost:8080",
+        mime = "application/json",
+    }
+    ```
+
+    Both snippets print:
+
+    ```
+    GET http://localhost:8080
+    Accept: application/json
+    ```
+
+    Note that `eprintdoc!` does not remove the final line, that's why we use `eprintln` instead of `eprint`.
+
+- `indoc!`: Dedents the passed string.
+
+    Before:
+
+    ```rs
+    indoc! {r#"
+        def hello():
+            print("Hello, world!")
+
+        hello()
+    "#}
+    ^^^^ // common leading whitespace (will be removed)
+    ```
+
+    With dedented string literals:
+
+    ```rs
+    dr#"
+        def hello():
+            print("Hello, world!")
+
+        hello()
+
+    "#
+    ^^^^ // common leading whitespace (will be removed)
+    ```
+
+    Both snippets evaluate to:
+
+    ```py
+    def hello():
+        print("Hello, world!")
+
+    hello()
+    ```
+
+    Note that `indoc!` does not remove the final line, that's why we add an additional newline after `hello()`.
+
+As a bonus, not only does it unify many macros under a single language feature.
+
+It also allows us to trivially create new macros that automatically make use of the feature in a backwards-compatible way.
+
+Take for instance the `text!` macro exported from the `iced` crate:
+
+```rs
+macro_rules! text {
+    ($($arg:tt)*) => {
+        $crate::Text::new(format!($($arg)*))
+    };
+}
+```
+
+In order to ergonomically supply a dedented string to it, one needs to re-create the macro:
+
+```rs
+macro_rules! textdoc {
+    ($($arg:tt)*) => {
+        iced::text!(formatdoc!($($arg)*))
+    };
+}
+```
+
+That's not a problem for *this* example, however with more involved macros such as [ones from the `log` crate](https://docs.rs/log/0.4.27/src/log/macros.rs.html#165-186) it becomes a problem. 
+
+With this RFC, re-implementing the macros is not going to be necessary anymore, as you can just pass in the dedented string literals:
+
+```rs
+text!(d"
+    GET {url}
+    Accept: {mime}
+")
+^^^^ // common leading whitespace (will be removed)
+```
+
+The language feature works with any user-defined macros that pass their arguments to `format_args!` under the hood.
+
+### Improved compile times
+
+Having dedented strings as a language feature could reduce compile time.
+
+- Users do not have to compile the crate *or* its dependencies.
+- There is no need for procedural macro expansion to take place in order to un-indent the macro. This step happens directly in the compiler.
```

Comment:
Worth mentioning that a builtin macro would probably have similar performance to the literal itself, so, this is only an issue for external crates.